### PR TITLE
Kpi sitewide

### PR DESCRIPTION
**Pull request by azizullahsaeidi**
1. Get top 10 viewed resources
2. Get total views
3. Get total registered users views
4. Get total guest users views

Please run below commands
```
php artisan migrate --path=/database/migrations/2024_05_01_124155_add_new_index_to_resource_views_table.php
```

```
php artisan migrate --path=/database/migrations/2024_05_02_181139_add_index_to_resources_table.php
```

```
php artisan migrate --path=/database/migrations/2024_07_01_060829_create_devices_table.php
```
```
php artisan migrate --path=/database/migrations/2024_07_01_062057_create_platforms_table.php
```

```
php artisan migrate --path=/database/migrations/2024_07_01_062632_create_browsers_table.php
```

```
php artisan migrate --path=/database/migrations/2024_07_01_065252_create_sitewide_page_views_table.php
```

```
php artisan migrate --path=/database/migrations/2024_07_03_154638_create_glossary_page_views_table.php
```

```
composer install
```

### CODE REVIEW

**Comment by azizullahsaeidi** [comment]
> Hi @azizullahsaeidi, thank you for working on this feature. I have a couple of questions before starting to review this:
> 
> 1. Why do we list 'top 10 resources' under the sitewide route? Shouldn't it be 'top 10 pages' instead?
> 2. What are we doing to eliminate bot views?
> 
> Thank you!

Thank you Aaron, yes you are right it should be top 10 pages.

**Comment by uyscuti-wiki** [comment]
Thank you @azizullahsaeidi for working on this. Few comments:

1. Are all the filters applicable for all the cards? If not, they should have indication as to which ones are applicable for which.
2. Please add some space between the two rows of the filter. They are too close. 
image
3. 'Page Type' should be 'Page type'. 'Is Rot' should be 'Is robot?'.
4. Let's rename all instances of 'sitewides' to its singular form - 'sitewide'. This includes the name of the migration as well. 
5. 'Total view by languages' should be 'Total views by languages'.
6. For 'top 10 viewed pages', let's hyperlink the URL with the URL itself. For example,  [https://library.darakhtdanesh.org/fa/resources](https://library.darakhtdanesh.org/fa/resources) instead of [Resource List](https://library.darakhtdanesh.org/fa/resources). 'Total Top 10 Viewed Pages' should be 'Total views'.
7. For 'total views based on user types', 'Total View' should be 'Total views', and not part of the numbered list, but at the end. 'Unregistered Users Views' should be 'Unregistered', and 'Registered Users Views' should be 'Registered'.
8. 'View based on platform' should be 'Views by platforms'.
9. 'View based on browser' should be 'Views by browsers'.
10. Let's drop the 'view based on device' card altogether.
11. Let's also drop the 'view based on page type' card.

New feature request:
1. Please add a new filter where, when an admin enters the URL of a page, a new card with the page views of that page is returned. ([DDL-163](https://ddlibrary.atlassian.net/browse/DDL-163))

**Comment by azizullahsaeidi** [comment]
> Thank you @azizullahsaeidi for working on this. Few comments:
> 
> 1. Are all the filters applicable for all the cards? If not, they should have indication as to which ones are applicable for which.
> 2. Please add some space between the two rows of the filter. They are too close.
>    image
> 3. 'Page Type' should be 'Page type'. 'Is Rot' should be 'Is robot?'.
> 4. Let's rename all instances of 'sitewides' to its singular form - 'sitewide'. This includes the name of the migration as well.
> 5. 'Total view by languages' should be 'Total views by languages'.
> 6. For 'top 10 viewed pages', let's hyperlink the URL with the URL itself. For example,  https://library.darakhtdanesh.org/fa/resources instead of [Resource List](https://library.darakhtdanesh.org/fa/resources). 'Total Top 10 Viewed Pages' should be 'Total views'.
> 7. For 'total views based on user types', 'Total View' should be 'Total views', and not part of the numbered list, but at the end. 'Unregistered Users Views' should be 'Unregistered', and 'Registered Users Views' should be 'Registered'.
> 8. 'View based on platform' should be 'Views by platforms'.
> 9. 'View based on browser' should be 'Views by browsers'.
> 10. Let's drop the 'view based on device' card altogether.
> 11. Let's also drop the 'view based on page type' card.
> 
> New feature request:
> 
> 1. Please add a new filter where, when an admin enters the URL of a page, a new card with the page views of that page is returned. ([DDL-163](https://ddlibrary.atlassian.net/browse/DDL-163))

Thank you Aaron

1. All the filters applicable for all the cards
2. Added more spaces
3. Changed 'Page Type' to 'Page type'. 'Is Rot' to 'Is bot'
4. Changed sitewides to sitewide 
5. 'Total view by languages' changed to 'Total views by languages'. 
6. Now displaying the link instead of the text
7. Changed `Total View` to `Total views`
8. Done
9. Done
10. Done
11. Done

File: database/migrations/2024_07_01_060809_create_page_types_table.php
Comment (uyscuti-wiki) [inline database/migrations/2024_07_01_060809_create_page_types_table.php:1]: Can we drop this?

File: database/migrations/2024_07_01_065252_create_sitewide_page_views_table.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    /**
+     * Run the migrations.
+     */
+    public function up(): void
+    {
+        Schema::create('sitewide_page_views', function (Blueprint $table) {
+            $table->id();
+            $table->string('page_url', 255);
+            $table->string('user_agent', 255);
+            $table->string('ip_address', 45);
```
Comment (uyscuti-wiki) [inline database/migrations/2024_07_01_065252_create_sitewide_page_views_table.php:17]: Drop this if possible.

File: database/migrations/2024_07_03_154638_create_glossary_page_views_table.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+return new class extends Migration {
+    /**
+     * Run the migrations.
+     */
+    public function up(): void
+    {
+        Schema::create('glossary_page_views', function (Blueprint $table) {
+            $table->id();
+            $table->string('user_agent', 255);
+            $table->string('ip_address', 45);
```
Comment (uyscuti-wiki) [inline database/migrations/2024_07_03_154638_create_glossary_page_views_table.php:16]: Remove IP if possible.

File: resources/views/admin/layouts/navigation.blade.php [hunk 3]
```diff
@@ -93,11 +93,13 @@
...
-        <a class="collapse-item" href="{{ URL::to('admin/reports/ga') }}">Google Analytics</a>
-        <a class="collapse-item" href="{{ URL::to('admin/reports/resources') }}">Resource Analytics</a>
+        <a class="collapse-item" href="{{ URL::to('admin/analytics/reports/glossary') }}">Glossary View</a>
+        <a class="collapse-item" href="{{ URL::to('admin/analytics/reports/sitewide') }}">Resource View</a>
         <div class="collapse-divider"></div>
         <h6 class="collapse-header">Analytics</h6>
+        <a class="collapse-item" href="{{ URL::to('admin/analytics/sitewide') }}">Sitewide Analytics</a>
         <a class="collapse-item" href="{{ URL::to('admin/analytics/resources') }}">Resource Analytics</a>
+        <a class="collapse-item" href="{{ URL::to('admin/analytics/glossaries') }}">Glossary Analytics</a>
         <a class="collapse-item" href="{{ URL::to('admin/analytics/users') }}">User Analytics</a>
```
Comment (uyscuti-wiki) [inline resources/views/admin/layouts/navigation.blade.php:103]: Let's make all routes singular - 'resource', 'glossary' and 'user'.

**Review by uyscuti-wiki — CHANGES_REQUESTED** [review]
Thank you Aziz for making the changes. Here are few more minor comments before we get this merged.

File: app/Traits/SitewidePageViewTrait.php [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            $device = Device::firstOrCreate(['name' => $agent->device()]);
+            $pageTypeModel = PageType::firstOrCreate(['name' => $pageType]);
+            $platform = Platform::firstOrCreate(['name' => $agent->platform()]);
+            $browser = Browser::firstOrCreate(['name' => $agent->browser    ()]);
+
+            SitewidePageView::insert([
+                'title' => $title,
+                'page_url' => $request->url(),
+                'user_agent' => $request->userAgent(),
+                'ip_address' => $request->ip(),
```
[the remark this answers is not in the document]
Comment (azizullahsaeidi) [reply app/Traits/SitewidePageViewTrait.php:34]: if you think IP address is not necessary so we can remove it.

File: app/Models/PageType.php
[the remark this answers is not in the document]
Comment (azizullahsaeidi) [reply app/Models/PageType.php:1]: I think we should keep page type. by page type we can know the number of visits of news, resources, contact-us, impact.

File: app/Http/Controllers/GlossaryAnalyticsController.php [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    private function getTotalViewBasedOnLanguage($request): Collection
+    {
+        $totalResources = GlossaryPageView::where(function ($query) use ($request) {
+            return $this->filterPageViews($query, $request);
+        })
+            ->groupBy('language')
+            ->select('language', DB::raw('count(*) as count'))
+            ->orderBy('count', 'desc')
```
[the remark this answers is not in the document]
Comment (azizullahsaeidi) [reply app/Http/Controllers/GlossaryAnalyticsController.php:92]: Done

File: app/Traits/SitewidePageViewTrait.php [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        $languageCode = $this->getLanguageCode();
+
+        if (count(explode("/$languageCode", $request->url())) == 2) {
+            $agent = new Agent();
+            $pageType = $this->getPageType($request->url());
+
+            $device = Device::firstOrCreate(['name' => $agent->device()]);
+            $pageTypeModel = PageType::firstOrCreate(['name' => $pageType]);
+            $platform = Platform::firstOrCreate(['name' => $agent->platform()]);
+            $browser = Browser::firstOrCreate(['name' => $agent->browser    ()]);
```
[the remark this answers is not in the document]
Comment (azizullahsaeidi) [reply app/Traits/SitewidePageViewTrait.php:28]: Done

File: resources/views/admin/layouts/navigation.blade.php (hunk 3 above)
Comment (azizullahsaeidi) [reply resources/views/admin/layouts/navigation.blade.php]: Done

File: app/Traits/SitewidePageViewTrait.php [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+        foreach ($patterns as $pattern => $type) {
+            if (preg_match("~{$pattern}~", $url)) {
+                return $type;
+            }
+        }
+
+        return 'page';
+    }
+
```
[the remark this answers is not in the document]
Comment (azizullahsaeidi) [reply app/Traits/SitewidePageViewTrait.php]: Removed

File: app/Traits/GlossaryPageViewTrait.php [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+            $device = Device::firstOrCreate(['name' => $agent->device()]);
+            $platform = Platform::firstOrCreate(['name' => $agent->platform()]);
+            $browser = Browser::firstOrCreate(['name' => $agent->browser()]);
+            $title = $request->subject ? ($status == GlossaryPageViewStatusEnum::Create ? $request->english : GlossarySubject::whereId($request->subject)->value('en')) : $title;
+
+            GlossaryPageView::insert([
+                'title' => $title,
+                'user_agent' => $request->userAgent(),
+                'ip_address' => $request->ip(),
```
[the remark this answers is not in the document]
Comment (azizullahsaeidi) [reply app/Traits/GlossaryPageViewTrait.php]: Removed

File: database/migrations/2024_07_01_060809_create_page_types_table.php
Comment (azizullahsaeidi) [reply database/migrations/2024_07_01_060809_create_page_types_table.php:1]: Removed

File: database/migrations/2024_07_01_065252_create_sitewide_page_views_table.php (hunk 1 above)
Comment (azizullahsaeidi) [reply database/migrations/2024_07_01_065252_create_sitewide_page_views_table.php]: ip_address removed

File: database/migrations/2024_07_03_154638_create_glossary_page_views_table.php (hunk 2 above)
Comment (azizullahsaeidi) [reply database/migrations/2024_07_03_154638_create_glossary_page_views_table.php]: Removed

**Review by uyscuti-wiki — APPROVED** [review]
Looks great. Thank you Aziz!

File: app/Http/Controllers/SitewideAnalyticsController.php [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+        $pageTypeCounts = PageType::select(['id', 'name'])
+            ->withCount([
+                'sitewidePageViews' => function ($query) use ($request) {
+                    return $this->filterPageViews($query, $request);
+                },
+            ])
+            ->get();
+
+        $totalGuestViews = $this->getTotalViews($request, 'yes');
+        $totalViewBasedOnLanguage = $this->getTotalViewBasedOnLanguage($request);
```
[the remark this answers is not in the document]
Comment (azizullahsaeidi) [reply app/Http/Controllers/SitewideAnalyticsController.php]: done

File: app/Http/Controllers/SitewideAnalyticsController.php [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+        return $query->count();
+    }
+
+    private function getTotalViewBasedOnLanguage($request): Collection
+    {
+        $totalResources = SitewidePageView::where(function ($query) use ($request) {
+            return $this->filterPageViews($query, $request);
+        })
+            ->groupBy('language')
+            ->select('language', DB::raw('count(*) as count'))
```
[the remark this answers is not in the document]
Comment (azizullahsaeidi) [reply app/Http/Controllers/SitewideAnalyticsController.php]: Done

**Pull request merged**